ceph/rados/core_workflows.py

```diff
@@ -870,24 +876,46 @@ def change_osd_state(self, action: str, target: int) -> bool:
             target: ID osd the target OSD
         Returns: Pass -> True, Fail -> False
         """
-
         cluster_fsid = self.run_ceph_command(cmd="ceph fsid")["fsid"]
-        osd_nodes = self.ceph_cluster.get_nodes(role="osd")
-        flag = False
-        for onode in osd_nodes:
-            res = self.collect_osd_daemon_ids(osd_node=onode)
-            if target in res:
-                cmd = f"systemctl {action} ceph-{cluster_fsid}@osd.{target}.service"
-                flag = True
-                log.info(
-                    f"Performing {action} on osd-{target} on host {onode.hostname}. Command {cmd}"
-                )
-                onode.exec_command(sudo=True, cmd=cmd)
-                # Sleeping for 5 seconds for status to be updated
-                time.sleep(5)
-                return True
-        if not flag:
-            log.error(f"osd-{target} not found on cluster")
+        host = self.fetch_host_node(daemon_type="osd", daemon_id=str(target))
+        if not host:
+            log.error("failed to find host for the osd")
+            return False
+        cmd = f"systemctl {action} ceph-{cluster_fsid}@osd.{target}.service"
+        log.info(
+            f"Performing {action} on osd-{target} on host {host.hostname}. Command {cmd}"
+        )
+        host.exec_command(sudo=True, cmd=cmd)
+        # Sleeping for 5 seconds for status to be updated
+        time.sleep(5)
+        return True
+
+    def fetch_host_node(self, daemon_type: str, daemon_id: str):
```

Comment:
Can we make it generic... for example `fetch_host_nodes` and have the argument `daemon_id` as optional?

Reply:
made changes